src/views/Rules/ListRuleRevisions/index.jsx

```diff
@@ -155,6 +200,17 @@ function ListRuleRevisions(props) {
           </Drawer>
         </Fragment>
       )}
+      <DialogAction
+        open={dialogState.open}
+        title={dialogState.title}
+        body={dialogState.body}
+        confirmText={dialogState.confirmText}
+        onSubmit={handleDialogSubmit}
+        onError={handleDialogError}
+        error={dialogState.error}
+        onComplete={handleDialogActionComplete}
+        onClose={handleDialogClose}
+      />
```

Comment:
When an error occurs, I am seeing an infinite spinner on the button.

![Screen Shot 2019-08-16 at 10 19 10 AM](https://user-images.githubusercontent.com/3766511/63174132-5aaf6f80-c00f-11e9-9c58-69a12ebb72c0.png)

I think it was working on the previous patch so it was probably something introduced after that.

Reply:
I'll look into this.

Reply:
Okay, so it looks like this is actually caused by the changes to `DialogAction` from #154. I did some more digging, and it appears that as things stand now, we set `cancelled = true` in every render - so we never `setActionExecuting(false)`.

I tried setting `[]` as the `useEffect` dependency again, which fixes this problem, but causes the initial error again. I think this is because setting `cancelled = true` happens in one render, and `setActionExecuting` happens in another (so `cancelled` ends up `false` again).

I also using a `setState` for `cancelled` instead of a local variable, but that doesn't appear to have helped. It seems like we need a value that persists across renders, is set to `true` when the component unmounts, and for `handleSubmit` to notice that change between calling `onComplete` and checking the value of `cancelled`. I'm a bit out of my depth here, so any tips would be appreciate.

Reply:
It looks like this is also causing problems when implementing the read only mutations for #152 - `setActionLoading` is never set to false, so subsequent opens of the dialog start the spinner already moving.

Reply:
I'll see how I can help. Worst case we revert the cleanup patch and fix that warning error later.

Reply:
Could you check if the following diff fixes it:

```diff
diff --git a/src/components/DialogAction/index.jsx b/src/components/DialogAction/index.jsx
index a6237a8..ab49127 100644
--- a/src/components/DialogAction/index.jsx
+++ b/src/components/DialogAction/index.jsx
@@ -1,4 +1,4 @@
-import React, { useEffect, useState } from 'react';
+import React, { useEffect, useRef, useState } from 'react';
 import { oneOfType, object, node, string, func, bool } from 'prop-types';
 import { makeStyles } from '@material-ui/styles';
 import Dialog from '@material-ui/core/Dialog';
@@ -30,6 +30,7 @@ const useStyles = makeStyles(theme => ({
 function DialogAction(props) {
   const classes = useStyles();
   const [actionExecuting, setActionExecuting] = useState(false);
+  const cancelled = useRef(false);
   const {
     open,
     title,
@@ -44,14 +45,17 @@ function DialogAction(props) {
     destructive,
     ...rest
   } = props;
-  let cancelled = false;
 
   // Set this so we can avoid calling setActionExecuting if the caller
   // decides to do something that unmounts us in onComplete or onError.
-  useEffect(() => () => (cancelled = true));
+  useEffect(() => () => {
+    cancelled.current = true;
+  });
 
   const handleSubmit = async () => {
-    setActionExecuting(true);
+    if (cancelled.current === false) {
+      setActionExecuting(true);
+    }
 
     const [error, result] = await tryCatch(onSubmit());
 
@@ -60,7 +64,7 @@ function DialogAction(props) {
         onError(error);
       }
 
-      if (!cancelled) {
+      if (cancelled.current === false) {
         setActionExecuting(false);
       }
 
@@ -71,7 +75,7 @@ function DialogAction(props) {
       onComplete(result);
     }
 
-    if (!cancelled) {
+    if (cancelled.current === false) {
       setActionExecuting(false);
     }
   };
```

^ uses the ref as an instance variable. (third point of https://github.com/mozilla-frontend-infra/balrog-ui/pull/142#discussion_r313188177)

Reply:
Ah, clearly I haven't interalized refs yet. Your patch works as long as I add an empty array as the second arg to `useEffect`.

Reply:
I tried removing the empty array and reverting or using the dialog elsewhere still seems to work. Any chance you could provide me with some STR of something that shows the dialog not working if the empty array is not specified?

Reply:
My STR is:
1) Load http://localhost:9000/rules/145/revisions
2) Click "Restore" on a revision
3) Click "Revert" on the dialog (oops, I need to change that text still).
4) Observe that while the button does change colour (light purple to dark purple), it does not get disabled.

With the empty array, both the Cancel and the Revert button get greyed out.